### PR TITLE
ci(feat): Workflow to prevent modified migrations

### PR DESCRIPTION
**Pull request by armenzg**
Modifying an existing migration can have very bad consequences and mess up the database setup of a developer and even production.
    
 We will allow modifying migrations but require new ones to make rectifications if that is what is intended.

### CODE REVIEW

**Comment by armenzg** [comment]
@billyvg I've placed the job under the backend lint workflow since it also runs on master (the other migration related workflows only run on pull requests.

Now, I realize that "backend lint" is *not* a required check. I believe this should be part of a required check before merging a PR.

Should we make one of the other two migration workflows to be required?
Should I move this job under another workflow that is required?
Should this be its own workflow that is required?

**Comment by billyvg** [comment]
IMO this workflow does not need to run in `master`, it serves its purpose on PRs and shouldn't allow the user to merge their PR.

We should add it to the `migrations.yml` workflow and make these both required.

**Comment by wedamija** [comment]
Is there a way to override this? For the most part, we never want to modify migrations (especially DDL related migrations). But for some backfills we do fix bugs in them over time. An example is 0024_auto_20191230_2052.py. This does some event migrations, and we periodically find bugs in this from self-hosted customers, which we fix to unblock them.

**Comment by armenzg** [comment]
> Is there a way to override this? For the most part, we never want to modify migrations (especially DDL related migrations). But for some backfills we do fix bugs in them over time. An example is 0024_auto_20191230_2052.py. This does some event migrations, and we periodically find bugs in this from self-hosted customers, which we fix to unblock them.

An admin on GH could override it and still merge the PR.

File: .github/workflows/migrations.yml
Comment (armenzg) [inline .github/workflows/migrations.yml:25]: @billyvg I don't think this is what you had in mind. Is there another way to prevent the downstream workflows to trigger?
image

**Comment by wedamija** [comment]
> > Is there a way to override this? For the most part, we never want to modify migrations (especially DDL related migrations). But for some backfills we do fix bugs in them over time. An example is 0024_auto_20191230_2052.py. This does some event migrations, and we periodically find bugs in this from self-hosted customers, which we fix to unblock them.
> 
> An admin on GH could override it and still merge the PR.

Any chance that we could make it so that members of @getsentry/owners-migrations could override?

**Comment by armenzg** [comment]
> > > Is there a way to override this? For the most part, we never want to modify migrations (especially DDL related migrations). But for some backfills we do fix bugs in them over time. An example is 0024_auto_20191230_2052.py. This does some event migrations, and we periodically find bugs in this from self-hosted customers, which we fix to unblock them.
> > 
> > 
> > An admin on GH could override it and still merge the PR.
> 
> Any chance that we could make it so that members of @getsentry/owners-migrations could override this check?

I can't get you that info as I don't have access. I believe an admin could let you know (or give us a screenshot) what options are available:
https://github.com/getsentry/sentry/settings/branches

**Comment by billyvg** [comment]
@wedamija If this is something that doesn't happen too often, can we stick to having an admin merge this in? I don't think there's a native way in GH to override a check and I'd prefer not to build a custom workflow for something that does not happen often. If it becomes painful, we can revisit it.

File: .github/workflows/migrations.yml
Comment (billyvg) [reply .github/workflows/migrations.yml:25]: @armenzg Similar to outputs for a step, [jobs also have `outputs`](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idoutputs) - we can then access them in dependent jobs with `needs.jobname.outputs`

Separate thing to note, `paths-filter` [provides a list of files](https://github.com/dorny/paths-filter#outputs) that was matched as an output

**Comment by armenzg** [comment]
It seems this is working.

With modified migration:
image

No migrations modified:
image
image

**Comment by wedamija** [comment]
> @wedamija If this is something that doesn't happen too often, can we stick to having an admin merge this in? I don't think there's a native way in GH to override a check and I'd prefer not to build a custom workflow for something that does not happen often. If it becomes painful, we can revisit it.

Yeah, that's ok with me. If possible, maybe have part of the error message say something like "If you have a valid reason to modify a migration please get approval from @getsentry/owners-migrations, then ask a github admin to merge".

File: .github/workflows/migrations.yml
Comment (billyvg) [inline .github/workflows/migrations.yml:46]: FYI, this is a breaking change between `v1` and `v2` - the former fetches ALL refs by default and `v2` only fetches a single commit. This may affect the `Get changed migration files` step.

However we could change that step to use the output of `steps.changes.outputs.migrations_added_files`

Comment (armenzg) [reply .github/workflows/migrations.yml]: Moving this to #24143

**Comment by armenzg** [comment]
@billyvg I believe we need to make this check as required; Would you please do it? Thanks!

**Review by billyvg — APPROVED** [review]

**Comment by billyvg** [comment]
@armenzg Let's wait a day or so before making it required, otherwise it can be a bit disruptive for out of date PRs.

**Pull request merged**
